# READ-ONLY as I will delete the model after 

<br><br><br>

In [1]:
# https://download.pytorch.org/models/vgg16_bn-6c64b313.pth

### This notebook is optionally accelerated with a GPU runtime.
### If you would like to use this acceleration, please select the menu option "Runtime" -> "Change runtime type", select "Hardware Accelerator" -> "GPU" and click "SAVE"

----------------------------------------------------------------------

# vgg-nets

*Author: Pytorch Team*

**Award winning ConvNets from 2014 Imagenet ILSVRC challenge**

<img src="https://pytorch.org/assets/images/vgg.png" alt="alt" width="50%"/>

In [2]:
import torch
# model = torch.hub.load('pytorch/vision:v0.6.0', 'vgg11', pretrained=True)
# or any of these variants
# model = torch.hub.load('pytorch/vision:v0.6.0', 'vgg11_bn', pretrained=True)
# model = torch.hub.load('pytorch/vision:v0.6.0', 'vgg13', pretrained=True)
# model = torch.hub.load('pytorch/vision:v0.6.0', 'vgg13_bn', pretrained=True)
# model = torch.hub.load('pytorch/vision:v0.6.0', 'vgg16', pretrained=True)
model = torch.hub.load('pytorch/vision:v0.6.0', 'vgg16_bn', pretrained=True)
# model = torch.hub.load('pytorch/vision:v0.6.0', 'vgg19', pretrained=True)
# model = torch.hub.load('pytorch/vision:v0.6.0', 'vgg19_bn', pretrained=True)
model.eval()
#  what does model.eval() do ??? 

Downloading: "https://github.com/pytorch/vision/archive/v0.6.0.zip" to C:\Users\tbresee/.cache\torch\hub\v0.6.0.zip
Downloading: "https://download.pytorch.org/models/vgg16_bn-6c64b313.pth" to C:\Users\tbresee/.cache\torch\hub\checkpoints\vgg16_bn-6c64b313.pth


VGG(
  (features): Sequential(
    (0): Conv2d(3, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (2): ReLU(inplace=True)
    (3): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (4): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (5): ReLU(inplace=True)
    (6): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
    (7): Conv2d(64, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (8): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (9): ReLU(inplace=True)
    (10): Conv2d(128, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (11): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (12): ReLU(inplace=True)
    (13): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
    (14): Conv2d(128, 256

In [5]:
model

VGG(
  (features): Sequential(
    (0): Conv2d(3, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (2): ReLU(inplace=True)
    (3): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (4): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (5): ReLU(inplace=True)
    (6): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
    (7): Conv2d(64, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (8): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (9): ReLU(inplace=True)
    (10): Conv2d(128, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (11): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (12): ReLU(inplace=True)
    (13): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
    (14): Conv2d(128, 256

# Examining the actual model ! 

In [25]:
model

VGG(
  (features): Sequential(
    (0): Conv2d(3, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (2): ReLU(inplace=True)
    (3): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (4): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (5): ReLU(inplace=True)
    (6): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
    (7): Conv2d(64, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (8): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (9): ReLU(inplace=True)
    (10): Conv2d(128, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (11): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (12): ReLU(inplace=True)
    (13): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
    (14): Conv2d(128, 256

In [7]:
dir(model)

['T_destination',
 '__annotations__',
 '__call__',
 '__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattr__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__setstate__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 '_apply',
 '_backward_hooks',
 '_buffers',
 '_call_impl',
 '_forward_hooks',
 '_forward_pre_hooks',
 '_get_name',
 '_initialize_weights',
 '_load_from_state_dict',
 '_load_state_dict_pre_hooks',
 '_modules',
 '_named_members',
 '_non_persistent_buffers_set',
 '_parameters',
 '_register_load_state_dict_pre_hook',
 '_register_state_dict_hook',
 '_replicate_for_data_parallel',
 '_save_to_state_dict',
 '_slow_forward',
 '_state_dict_hooks',
 '_version',
 'add_module',
 'apply',
 'avgpool',
 'bfloat16',
 'buffers',
 'children',
 'classifier',
 

In [8]:
model.features

Sequential(
  (0): Conv2d(3, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
  (1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (2): ReLU(inplace=True)
  (3): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
  (4): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (5): ReLU(inplace=True)
  (6): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
  (7): Conv2d(64, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
  (8): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (9): ReLU(inplace=True)
  (10): Conv2d(128, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
  (11): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (12): ReLU(inplace=True)
  (13): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
  (14): Conv2d(128, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 

In [10]:
model.features

Sequential(
  (0): Conv2d(3, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
  (1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (2): ReLU(inplace=True)
  (3): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
  (4): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (5): ReLU(inplace=True)
  (6): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
  (7): Conv2d(64, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
  (8): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (9): ReLU(inplace=True)
  (10): Conv2d(128, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
  (11): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (12): ReLU(inplace=True)
  (13): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
  (14): Conv2d(128, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 

In [13]:
model.type

<bound method Module.type of VGG(
  (features): Sequential(
    (0): Conv2d(3, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (2): ReLU(inplace=True)
    (3): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (4): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (5): ReLU(inplace=True)
    (6): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
    (7): Conv2d(64, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (8): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (9): ReLU(inplace=True)
    (10): Conv2d(128, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (11): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (12): ReLU(inplace=True)
    (13): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=Fal

In [18]:
# same as next  list(model.parameters())

In [17]:
for parameter in model.parameters():
    print(parameter)

Parameter containing:
tensor([[[[ 8.2833e-02,  2.7968e-02,  7.7096e-02],
          [ 4.9341e-02, -3.3441e-02,  1.9572e-02],
          [ 8.0300e-02,  7.7076e-02,  8.3349e-02]],

         [[-4.4296e-02, -1.7748e-01, -4.8706e-02],
          [-1.1003e-01, -2.7530e-01, -1.3474e-01],
          [-5.9982e-03, -6.1375e-02,  1.6822e-02]],

         [[ 2.7480e-02, -6.6769e-02,  4.3955e-02],
          [-2.6662e-02, -1.4995e-01, -3.3615e-02],
          [ 5.2778e-02,  1.7143e-02,  8.6744e-02]]],


        [[[-1.2628e-02,  3.0218e-02, -2.6930e-02],
          [-1.3764e-02,  1.1993e-01, -6.6263e-03],
          [-2.6019e-02, -8.3535e-03, -3.9197e-02]],

         [[-4.0557e-02,  1.3983e-02, -5.4278e-02],
          [ 1.5412e-02,  1.8198e-01,  1.7598e-02],
          [-1.7032e-02,  1.1284e-02, -2.4226e-02]],

         [[-6.5683e-02,  5.9252e-02, -5.3020e-02],
          [ 3.8278e-02,  2.7292e-01,  5.9491e-02],
          [-4.1218e-02,  3.6159e-02, -3.0478e-02]]],


        [[[ 1.4962e-06, -1.1430e-06,  1.2536

         3.0883e-01], device='cuda:0', requires_grad=True)
Parameter containing:
tensor([ 6.1194e-02, -2.0338e-02,  6.9390e-02, -1.6764e-02, -1.9737e-01,
         1.9742e-01, -1.0016e-01, -1.7291e-01, -1.6403e-01, -1.6123e-01,
        -1.6141e-01,  2.5101e-01, -7.2010e-02,  2.4077e-01, -1.2702e-01,
        -9.8436e-02, -1.2555e-01,  2.0554e-03, -1.3933e-01,  3.5958e-02,
        -1.1799e-01,  2.6867e-01,  2.9129e-02, -8.0782e-02, -1.7855e-01,
        -1.4094e-01, -8.5141e-02, -4.9177e-02, -1.0380e-01, -1.8253e-01,
        -1.1369e-01, -1.9760e-01, -7.3630e-02, -4.1572e-02, -1.7843e-01,
        -8.8795e-02, -1.3544e-01, -3.0692e-03, -8.6868e-02,  6.2045e-02,
        -7.4820e-02, -8.1671e-02, -2.2117e-01, -6.6426e-02, -1.0768e-01,
        -1.3577e-01,  3.6913e-01, -1.2038e-01, -5.9190e-02, -1.3874e-01,
         1.2675e-01,  9.8803e-02, -1.5123e-01, -4.5077e-02, -1.3915e-01,
        -6.6811e-02, -1.1065e-02,  2.3155e-01, -6.5496e-02, -3.3038e-01,
        -1.1347e-01,  4.8300e-02, -5.5851e-

         2.4308e-08, -1.7686e-08], device='cuda:0', requires_grad=True)
Parameter containing:
tensor([0.2935, 0.2028, 0.2278, 0.2812, 0.2415, 0.2887, 0.2899, 0.2003, 0.2941,
        0.2560, 0.3168, 0.2435, 0.3322, 0.3298, 0.2133, 0.2591, 0.2522, 0.1981,
        0.2415, 0.2533, 0.2305, 0.2925, 0.2994, 0.2615, 0.2720, 0.2227, 0.2926,
        0.2413, 0.1946, 0.2358, 0.2498, 0.2444, 0.2506, 0.2319, 0.1841, 0.3238,
        0.2379, 0.2455, 0.2290, 0.3200, 0.2885, 0.2446, 0.3003, 0.3444, 0.2188,
        0.2233, 0.2232, 0.2345, 0.2844, 0.2483, 0.3061, 0.2620, 0.2760, 0.1997,
        0.2439, 0.2314, 0.2240, 0.2164, 0.2549, 0.2742, 0.3209, 0.2169, 0.2534,
        0.2281, 0.1678, 0.2688, 0.2606, 0.2355, 0.3214, 0.1893, 0.2665, 0.2737,
        0.2830, 0.2774, 0.3148, 0.3267, 0.1472, 0.2263, 0.2715, 0.2272, 0.2736,
        0.2581, 0.1928, 0.2443, 0.2618, 0.2663, 0.2470, 0.2491, 0.2880, 0.2165,
        0.2620, 0.2152, 0.2795, 0.2302, 0.2643, 0.2239, 0.2631, 0.2446, 0.2493,
        0.2614, 0.2863, 0.

        -3.1086e-08, -3.9289e-08], device='cuda:0', requires_grad=True)
Parameter containing:
tensor([0.2314, 0.2987, 0.2840, 0.2335, 0.2621, 0.2540, 0.2565, 0.2455, 0.2250,
        0.2473, 0.1974, 0.2187, 0.2185, 0.2400, 0.2635, 0.2443, 0.2232, 0.3169,
        0.2822, 0.2278, 0.2299, 0.2778, 0.2130, 0.2544, 0.2467, 0.2795, 0.2382,
        0.2959, 0.4218, 0.2217, 0.2681, 0.1731, 0.0387, 0.2790, 0.2559, 0.2271,
        0.2293, 0.2458, 0.2582, 0.2026, 0.2423, 0.3231, 0.2022, 0.2316, 0.2770,
        0.1774, 0.2191, 0.2476, 0.2458, 0.2353, 0.2815, 0.2527, 0.2606, 0.2749,
        0.2705, 0.2567, 0.3056, 0.2627, 0.2231, 0.0706, 0.2229, 0.2222, 0.2198,
        0.2799, 0.2595, 0.2697, 0.2502, 0.2159, 0.2658, 0.2868, 0.2843, 0.3030,
        0.2714, 0.2495, 0.3081, 0.3435, 0.2124, 0.2504, 0.2305, 0.2424, 0.2115,
        0.1795, 0.2134, 0.2832, 0.2598, 0.3347, 0.1975, 0.2076, 0.2839, 0.1729,
        0.2393, 0.2612, 0.2479, 0.2373, 0.2457, 0.2687, 0.2350, 0.2096, 0.2380,
        0.2359, 0.2115, 0.

        -7.4101e-09, -6.9329e-08], device='cuda:0', requires_grad=True)
Parameter containing:
tensor([0.3018, 0.2847, 0.3348, 0.2648, 0.2303, 0.3118, 0.3544, 0.3043, 0.2952,
        0.3515, 0.3786, 0.3209, 0.2653, 0.2347, 0.2602, 0.3044, 0.2762, 0.2976,
        0.2345, 0.2738, 0.3097, 0.2150, 0.3184, 0.3773, 0.2642, 0.1980, 0.2262,
        0.2711, 0.2981, 0.2418, 0.2810, 0.2341, 0.3082, 0.2675, 0.2378, 0.2419,
        0.2508, 0.2466, 0.2794, 0.4442, 0.2871, 0.2972, 0.3052, 0.3650, 0.2915,
        0.1944, 0.3310, 0.2551, 0.2208, 0.3038, 0.2035, 0.3116, 0.2995, 0.2100,
        0.3326, 0.2982, 0.3029, 0.2616, 0.2592, 0.2819, 0.3019, 0.3024, 0.3110,
        0.2804, 0.3285, 0.2188, 0.2741, 0.2994, 0.2917, 0.2720, 0.2972, 0.2318,
        0.2741, 0.1924, 0.2898, 0.2704, 0.3065, 0.5719, 0.3151, 0.2628, 0.3605,
        0.2908, 0.2072, 0.2574, 0.2751, 0.2774, 0.2954, 0.2382, 0.2289, 0.2687,
        0.3480, 0.3682, 0.2560, 0.2185, 0.2680, 0.2844, 0.3142, 0.3034, 0.3199,
        0.2668, 0.1827, 0.

Parameter containing:
tensor([-1.3794e-02, -1.7240e-02, -2.7464e-02, -1.9545e-02,  8.8474e-03,
        -1.7590e-02, -2.4578e-02,  1.1173e-03, -1.1956e-02, -2.3097e-02,
        -8.2994e-03, -4.0352e-03,  1.0268e-02,  1.8215e-03,  2.3942e-03,
        -1.0785e-02,  1.5384e-02, -6.5712e-03,  2.7475e-03,  2.6030e-03,
        -2.8478e-02,  1.7517e-02, -6.2749e-03, -7.7598e-03, -1.9992e-02,
        -6.2223e-03, -1.1509e-02, -8.7595e-03, -1.9052e-03, -1.2204e-02,
         1.3407e-03, -3.3139e-03, -9.9430e-04, -8.1035e-03, -8.0503e-03,
        -2.4488e-02, -4.3193e-03,  3.0826e-02, -2.1367e-02,  9.1784e-03,
         2.2933e-02,  1.0686e-02,  5.5708e-04, -2.4824e-02,  1.4926e-02,
        -6.0772e-03,  4.0836e-02, -1.6081e-02, -4.5205e-03, -1.2169e-02,
        -6.8472e-03, -8.9579e-03, -1.2841e-02,  2.0208e-02, -1.5336e-02,
        -3.4192e-03,  1.2269e-04,  5.4123e-02,  2.8981e-02,  2.2311e-02,
        -5.1162e-02,  3.0154e-04,  6.7782e-03,  2.3118e-02,  3.7043e-02,
        -1.7705e-02, -2.9856e

In [19]:
def count_parameters(model):
    return sum(p.numel() for p in model.parameters() if p.requires_grad)
count_parameters(model)

138365992

In [20]:
import torch
from torchvision import models
from torchsummary import summary

device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
vgg = models.vgg16().to(device)

summary(vgg, (3, 224, 224))  # always input will be form of 3 layers rgb with 224, 

----------------------------------------------------------------
        Layer (type)               Output Shape         Param #
            Conv2d-1         [-1, 64, 224, 224]           1,792
              ReLU-2         [-1, 64, 224, 224]               0
            Conv2d-3         [-1, 64, 224, 224]          36,928
              ReLU-4         [-1, 64, 224, 224]               0
         MaxPool2d-5         [-1, 64, 112, 112]               0
            Conv2d-6        [-1, 128, 112, 112]          73,856
              ReLU-7        [-1, 128, 112, 112]               0
            Conv2d-8        [-1, 128, 112, 112]         147,584
              ReLU-9        [-1, 128, 112, 112]               0
        MaxPool2d-10          [-1, 128, 56, 56]               0
           Conv2d-11          [-1, 256, 56, 56]         295,168
             ReLU-12          [-1, 256, 56, 56]               0
           Conv2d-13          [-1, 256, 56, 56]         590,080
             ReLU-14          [-1, 256,

In [32]:

for name, param in model.named_parameters(): 
    if param.requires_grad: 
        print ("\n%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%:\n",name, "\n\n", param.data)



%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%:
 features.0.weight 

 tensor([[[[ 8.2833e-02,  2.7968e-02,  7.7096e-02],
          [ 4.9341e-02, -3.3441e-02,  1.9572e-02],
          [ 8.0300e-02,  7.7076e-02,  8.3349e-02]],

         [[-4.4296e-02, -1.7748e-01, -4.8706e-02],
          [-1.1003e-01, -2.7530e-01, -1.3474e-01],
          [-5.9982e-03, -6.1375e-02,  1.6822e-02]],

         [[ 2.7480e-02, -6.6769e-02,  4.3955e-02],
          [-2.6662e-02, -1.4995e-01, -3.3615e-02],
          [ 5.2778e-02,  1.7143e-02,  8.6744e-02]]],


        [[[-1.2628e-02,  3.0218e-02, -2.6930e-02],
          [-1.3764e-02,  1.1993e-01, -6.6263e-03],
          [-2.6019e-02, -8.3535e-03, -3.9197e-02]],

         [[-4.0557e-02,  1.3983e-02, -5.4278e-02],
          [ 1.5412e-02,  1.8198e-01,  1.7598e-02],
          [-1.7032e-02,  1.1284e-02, -2.4226e-02]],

         [[-6.5683e-02,  5.9252e-02, -5.3020e-02],
          [ 3.8278e-02,  2.7292e-01,  5.9491e-02],
          [-4.1218e-02,  3.6159e-02, -3.0478e-02]]],


        [

 features.14.bias 

 tensor([-1.6105e-08,  7.5244e-08,  2.1525e-08, -7.4830e-08,  4.0028e-08,
         2.0125e-07,  3.0836e-09, -1.2907e-07, -4.8796e-09, -5.7191e-08,
        -5.4578e-08, -3.9656e-08, -2.1347e-09, -1.1581e-07,  7.6258e-08,
        -2.0476e-08,  1.6960e-08,  1.0894e-07,  7.6891e-08,  3.2939e-08,
        -2.4267e-09, -6.8908e-08,  8.4935e-08, -4.5520e-08,  8.2543e-08,
        -5.2748e-08, -9.7704e-08,  3.9908e-10,  6.0350e-09,  2.8966e-08,
        -1.8077e-08,  4.6512e-08, -4.5208e-08, -5.3004e-08,  3.0517e-08,
         8.5009e-09, -3.7000e-08,  2.4454e-08, -5.9350e-08, -1.2664e-08,
        -2.0602e-08,  8.7701e-09, -5.6419e-08, -2.7916e-08, -3.2290e-08,
        -4.0166e-08, -5.9390e-09,  2.5727e-08,  6.1466e-09,  4.4333e-08,
        -1.0324e-07, -2.8709e-08,  4.4208e-08, -7.7599e-08,  8.6090e-09,
        -1.9044e-08,  1.1488e-07,  1.1980e-07,  5.7397e-08,  3.7494e-08,
         5.3562e-08, -6.0933e-08,  8.5125e-09,  9.2560e-08, -5.4153e-08,
         1.5346e-08, -2.5016e-

          [-2.0245e-02,  5.5722e-04,  1.2079e-02]]]], device='cuda:0')

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%:
 features.24.bias 

 tensor([-1.1572e-08,  3.5092e-08,  1.7648e-08, -8.9680e-09, -2.3482e-09,
        -5.8172e-09,  4.3285e-09,  5.0526e-09, -5.8743e-09,  4.8384e-09,
         7.8931e-08, -8.7426e-09,  1.8629e-08,  1.0511e-08,  2.0101e-09,
         1.5243e-09,  4.4172e-08,  4.3102e-09, -2.4948e-08, -1.0873e-08,
        -2.2158e-08,  7.0426e-10,  2.3177e-08,  3.9670e-08, -4.4207e-08,
        -1.9966e-08,  2.5604e-08, -1.5335e-08,  9.8608e-09,  2.9854e-08,
         5.6739e-09,  3.4812e-08, -4.7872e-08,  4.2728e-08, -2.8230e-08,
        -2.8395e-08,  3.4833e-09,  2.4321e-08, -2.3082e-08,  2.7358e-08,
         5.8142e-08, -8.2889e-08, -2.9753e-08,  8.4307e-09,  7.1400e-08,
        -3.0681e-09, -1.5972e-08, -2.4394e-08, -5.9544e-08,  3.9597e-08,
        -3.2129e-09,  1.1697e-08, -5.7131e-08, -1.7815e-08,  2.8140e-08,
         1.8085e-08, -1.6909e-08, -2.3169e-08,  1.3807e-08, -6.8456e-0

          [ 8.1887e-03, -8.2345e-03,  9.0428e-03]]]], device='cuda:0')

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%:
 features.30.bias 

 tensor([ 1.9902e-08, -8.2983e-08, -1.4225e-08, -9.5289e-10, -2.0318e-08,
         2.7256e-08, -2.8104e-08, -2.4539e-08,  7.5508e-08, -2.7550e-09,
         2.8892e-09,  2.3473e-10, -1.5730e-08, -6.8345e-08,  3.7255e-08,
         1.7291e-10,  5.3689e-08,  8.2421e-09, -5.2291e-08, -3.9778e-08,
         1.0463e-07, -3.9784e-08, -3.1304e-08, -7.2300e-08,  5.3896e-09,
         8.8980e-09, -6.7317e-08, -6.6899e-08, -5.3453e-08, -4.0407e-08,
         1.6766e-08, -1.0262e-08, -5.8942e-09, -6.7277e-08, -6.8004e-08,
        -2.5771e-08, -4.5430e-08,  1.2334e-08,  7.6341e-08,  3.7965e-08,
        -4.0646e-08, -5.7450e-08,  2.9834e-08, -6.8584e-08, -7.3996e-08,
        -4.4447e-08, -7.7803e-08, -3.6172e-08, -6.7825e-08, -1.2261e-08,
        -3.8342e-08, -4.0322e-08, -4.4039e-08, -6.8079e-08,  6.3074e-09,
        -4.3635e-08,  8.1406e-09,  6.6266e-08, -4.3571e-08,  3.2880e-0

          [-1.3112e-02, -1.1173e-02, -1.3277e-02]]]], device='cuda:0')

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%:
 features.37.bias 

 tensor([-7.2728e-08, -5.0621e-08, -8.6599e-08, -1.8289e-08, -2.7556e-08,
        -4.9061e-08, -7.1937e-08, -2.4942e-08,  1.0501e-08, -4.7339e-08,
         1.6268e-08,  1.3585e-08, -1.7072e-08, -4.1578e-09, -3.9241e-08,
        -1.1183e-08, -4.5178e-08, -3.5511e-08, -6.0744e-08, -1.2534e-08,
        -8.8289e-08,  3.9879e-08, -4.7682e-08, -6.3695e-08, -1.6259e-09,
        -2.1735e-08,  7.9524e-09, -2.2138e-08, -3.2543e-08, -4.4767e-08,
        -2.0012e-08, -3.5980e-08, -8.0691e-08,  1.6015e-09, -5.0162e-08,
        -8.8592e-09, -4.9377e-08, -6.9195e-08, -3.2567e-08,  1.7141e-08,
        -7.4960e-08, -6.9869e-08,  2.2559e-08, -8.3603e-09, -2.3324e-08,
         4.6377e-09, -3.8382e-08, -7.7284e-08, -7.0382e-10, -5.1563e-08,
         2.4686e-08, -3.2290e-08, -4.5045e-08,  1.7725e-08, -1.0256e-07,
        -8.4036e-08, -4.4846e-08, -3.7378e-08, -3.5916e-08, -6.0990e-0

       device='cuda:0')

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%:
 classifier.3.bias 

 tensor([0.0289, 0.1051, 0.1270,  ..., 0.1079, 0.0733, 0.0951], device='cuda:0')

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%:
 classifier.6.weight 

 tensor([[-7.1171e-04,  3.1785e-02, -1.8842e-02,  ..., -1.5870e-03,
          4.0962e-02,  1.9549e-03],
        [-1.7012e-02,  4.7915e-02, -1.6017e-02,  ..., -1.1342e-02,
          3.7112e-02, -2.4083e-06],
        [ 1.9150e-03, -1.3026e-02, -6.5281e-03,  ..., -1.3581e-02,
         -2.7002e-02,  4.0695e-02],
        ...,
        [ 2.2554e-02,  2.1627e-03, -5.9601e-03,  ..., -2.0709e-02,
         -8.1781e-03, -1.1481e-02],
        [-8.4101e-03,  2.8122e-03, -1.0132e-02,  ...,  1.6001e-02,
         -4.0399e-03,  3.7749e-02],
        [-1.4075e-02, -3.9555e-03,  2.6998e-02,  ..., -1.8265e-02,
          3.6453e-03, -1.4508e-02]], device='cuda:0')

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%:
 classifier.6.bias 

 tensor([-1.3794e-02, -1.7240e-02, -2.7464e-02, -1.9545e-02,  8.8474e-03,
 

##### important:

In [24]:
for name, param in model.named_parameters():
    print(name)

features.0.weight
features.0.bias
features.1.weight
features.1.bias
features.3.weight
features.3.bias
features.4.weight
features.4.bias
features.7.weight
features.7.bias
features.8.weight
features.8.bias
features.10.weight
features.10.bias
features.11.weight
features.11.bias
features.14.weight
features.14.bias
features.15.weight
features.15.bias
features.17.weight
features.17.bias
features.18.weight
features.18.bias
features.20.weight
features.20.bias
features.21.weight
features.21.bias
features.24.weight
features.24.bias
features.25.weight
features.25.bias
features.27.weight
features.27.bias
features.28.weight
features.28.bias
features.30.weight
features.30.bias
features.31.weight
features.31.bias
features.34.weight
features.34.bias
features.35.weight
features.35.bias
features.37.weight
features.37.bias
features.38.weight
features.38.bias
features.40.weight
features.40.bias
features.41.weight
features.41.bias
classifier.0.weight
classifier.0.bias
classifier.3.weight
classifier.3.bias


---

---

```
VGG(
  (features): Sequential(
    (0): Conv2d(3, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (2): ReLU(inplace=True)
    (3): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (4): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (5): ReLU(inplace=True)
    (6): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
    (7): Conv2d(64, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (8): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (9): ReLU(inplace=True)
    (10): Conv2d(128, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (11): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (12): ReLU(inplace=True)
    (13): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
    (14): Conv2d(128, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (15): BatchNorm2d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (16): ReLU(inplace=True)
    (17): Conv2d(256, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (18): BatchNorm2d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (19): ReLU(inplace=True)
    (20): Conv2d(256, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (21): BatchNorm2d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (22): ReLU(inplace=True)
    (23): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
    (24): Conv2d(256, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (25): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (26): ReLU(inplace=True)
    (27): Conv2d(512, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (28): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (29): ReLU(inplace=True)
    (30): Conv2d(512, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (31): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (32): ReLU(inplace=True)
    (33): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
    (34): Conv2d(512, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (35): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (36): ReLU(inplace=True)
    (37): Conv2d(512, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (38): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (39): ReLU(inplace=True)
    (40): Conv2d(512, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (41): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (42): ReLU(inplace=True)
    (43): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
  )
  (avgpool): AdaptiveAvgPool2d(output_size=(7, 7))
  (classifier): Sequential(
    (0): Linear(in_features=25088, out_features=4096, bias=True)
    (1): ReLU(inplace=True)
    (2): Dropout(p=0.5, inplace=False)
    (3): Linear(in_features=4096, out_features=4096, bias=True)
    (4): ReLU(inplace=True)
    (5): Dropout(p=0.5, inplace=False)
    (6): Linear(in_features=4096, out_features=1000, bias=True)
  )
)
```

All pre-trained models expect input images normalized in the same way,
i.e. mini-batches of 3-channel RGB images of shape `(3 x H x W)`, where `H` and `W` are expected to be at least `224`.
The images have to be loaded in to a range of `[0, 1]` and then normalized using `mean = [0.485, 0.456, 0.406]`
and `std = [0.229, 0.224, 0.225]`.

Here's a sample execution.

In [3]:
# Download an example image from the pytorch website
import urllib
url, filename = ("https://github.com/pytorch/hub/raw/master/images/dog.jpg", "dog.jpg")
try: urllib.URLopener().retrieve(url, filename)
except: urllib.request.urlretrieve(url, filename)

In [4]:
# sample execution (requires torchvision)
from PIL import Image
from torchvision import transforms
input_image = Image.open(filename)
preprocess = transforms.Compose([
    transforms.Resize(256),
    transforms.CenterCrop(224),
    transforms.ToTensor(),
    transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]),
])
input_tensor = preprocess(input_image)
input_batch = input_tensor.unsqueeze(0) # create a mini-batch as expected by the model

# move the input and model to GPU for speed if available
if torch.cuda.is_available():
    input_batch = input_batch.to('cuda')
    model.to('cuda')

with torch.no_grad():
    output = model(input_batch)
# Tensor of shape 1000, with confidence scores over Imagenet's 1000 classes
print(output[0])
# The output has unnormalized scores. To get probabilities, you can run a softmax on it.
print(torch.nn.functional.softmax(output[0], dim=0))


tensor([-8.8345e-01, -1.0631e+00, -1.2372e+00, -2.4203e+00, -2.0970e+00,
        -8.2611e-01, -1.3973e+00,  3.0659e+00,  5.0388e+00, -1.5407e+00,
        -4.0731e+00, -1.9172e+00, -2.7595e+00, -1.9206e+00, -3.0181e+00,
        -2.1254e+00, -1.0687e+00,  6.4184e-01, -3.4431e-01, -1.8556e+00,
        -2.0571e+00, -1.6285e+00, -2.8914e-01, -2.6233e-01, -2.9559e+00,
        -2.7580e+00, -3.1451e+00, -1.9105e+00, -1.3095e+00,  1.9597e+00,
        -2.7696e+00, -3.0991e+00, -2.9593e+00, -3.4072e+00, -3.7172e+00,
        -3.9573e+00, -3.1031e+00, -4.2786e+00, -3.7760e+00, -1.2602e+00,
        -1.9135e+00, -3.6816e+00, -4.4522e+00, -4.4409e+00, -1.3955e+00,
        -3.2293e+00, -1.2520e+00, -3.8684e+00, -5.0446e+00, -1.5712e-01,
        -6.9244e-02, -2.6021e+00, -1.4988e+00, -2.8035e+00, -2.9700e+00,
        -2.4441e+00, -2.1299e+00, -3.2342e+00, -3.0140e+00, -2.7280e+00,
        -1.8944e+00, -3.5194e+00, -2.7883e+00, -4.3590e+00, -3.0985e+00,
        -2.3116e+00, -2.6788e+00, -2.3383e+00, -3.4

# More 

### Model Description

Here we have implementations for the models proposed in [Very Deep Convolutional Networks for Large-Scale Image Recognition](https://arxiv.org/abs/1409.1556),
for each configurations and their with bachnorm version.

For example, configuration `A` presented in the paper is `vgg11`, configuration `B` is `vgg13`, configuration `D` is `vgg16`
and configuration `E` is `vgg19`. Their batchnorm version are suffixed with `_bn`.

Their 1-crop error rates on imagenet dataset with pretrained models are listed below.

| Model structure | Top-1 error | Top-5 error |
| --------------- | ----------- | ----------- |
|  vgg11          | 30.98       | 11.37       |
|  vgg11_bn       | 26.70       | 8.58        |
|  vgg13          | 30.07       | 10.75       |
|  vgg13_bn       | 28.45       | 9.63        |
|  vgg16          | 28.41       | 9.62        |
|  vgg16_bn       | 26.63       | 8.50        |
|  vgg19          | 27.62       | 9.12        |
|  vgg19_bn       | 25.76       | 8.15        |

### References

- [Very Deep Convolutional Networks for Large-Scale Image Recognition](https://arxiv.org/abs/1409.1556).